File: c/src/intent_engine.c
#include "intent_engine.h"
#include <stdlib.h>
#include <string.h>

static IntentAspect aspect_from_input_type(const char* type) {
    if (!type) return INTENT_ASPECT_HARMONIC;
    if (strcmp(type, "chord_change") == 0 || strcmp(type, "harmonic") == 0) return INTENT_ASPECT_HARMONIC;
    if (strcmp(type, "rhythmic") == 0) return INTENT_ASPECT_RHYTHMIC;
    if (strcmp(type, "dynamic") == 0) return INTENT_ASPECT_DYNAMIC;
    if (strcmp(type, "tempo") == 0) return INTENT_ASPECT_TEMPO;
    return INTENT_ASPECT_HARMONIC;
}

MusicalIntent* infer_intent(const AbstractInput* input) {
    if (!input) return musical_intent_create();
    IntentAspect aspect = input->aspect;
    if (input->type && strlen(input->type) > 0) {
        aspect = aspect_from_input_type(input->type);
    }
    IntentComponent comp = {aspect, input->magnitude, "inferred"};
    return musical_intent_create_with_components(&comp, 1);
}

void derive_value_states(const MusicalIntent* intent, const ConstraintSet* constraints,
                         ValueStateSet* out_states, ViolationList* out_violations) {
    if (!intent || !out_states || !out_violations) return;
    *out_states = value_state_set_create();
    for (int i = 0; i < VALUE_TYPE_COUNT; i++) {
        ValueType vt = (ValueType)i;
        IntentAspect aspect = INTENT_ASPECT_HARMONIC;
        switch (vt) {
            case VALUE_TYPE_HARMONIC_TENSION: aspect = INTENT_ASPECT_HARMONIC; break;
            case VALUE_TYPE_RHYTHMIC_DENSITY: aspect = INTENT_ASPECT_RHYTHMIC; break;
            case VALUE_TYPE_DYNAMIC_RANGE: aspect = INTENT_ASPECT_DYNAMIC; break;
            case VALUE_TYPE_TEMPO_BIAS: aspect = INTENT_ASPECT_TEMPO; break;
            default: break;
        }
        bool has_intent = musical_intent_has_aspect(intent, aspect);
        bool constrained = false;
        if (constraints) {
            for (size_t j = 0; j < constraints->count; j++) {
                if (constraint_overrides_intent(&constraints->constraints[j], vt)) {
                    constrained = true;
                    break;
                }
            }
        }
        if (!has_intent) {
            value_state_set_set(out_states, vt, value_state_missing("no_intent"));
        } else {
            double mag = musical_intent_get_magnitude(intent, aspect);
            ValueState vs = value_state_derive(vt, mag, constrained);
            value_state_set_set(out_states, vt, vs);
        }
    }
    if (constraints) {
        constraints_apply(constraints, out_states, out_violations);
    }
}

bool validate_invariants(const IntentResult* result) {
    if (!result) return false;
    if (!result->intent) return false;
    if (!result->frozen) return false;
    if (!harmonic_escalation_satisfies_invariant(&result->harmonic_escalation)) return false;
    if (!rhythmic_escalation_satisfies_invariant(&result->rhythmic_escalation)) return false;
    for (int i = 0; i < VALUE_TYPE_COUNT; i++) {
        if (!value_state_validate(&result->value_states.states[i], (ValueType)i)) return false;
    }
    return true;
}

static bool is_empty_input(const AbstractInput* input) {
    if (!input) return true;
    if (input->type && input->type[0] != '\0') return false;
    return true;
}

/* Escalation v0: "more_harmonic_color" only. Two mutually exclusive branches: 7ths OR one inversion. */
static void compute_harmonic_escalation(const AbstractInput* input, const MusicalIntent* intent,
                                        HarmonicEscalationBranches* out) {
    harmonic_escalation_branches_clear(out);
    if (!input || !intent || !out) return;
    if (!input->escalation_token || strcmp(input->escalation_token, "more_harmonic_color") != 0)
        return;
    if (!musical_intent_has_aspect(intent, INTENT_ASPECT_HARMONIC))
        return;
    harmonic_escalation_branches_set_more_harmonic_color(out);
}

/* Escalation v1: "more_groove" only. Three branches: forward push, backward layback, neutral. */
static void compute_rhythmic_escalation(const AbstractInput* input, const MusicalIntent* intent,
                                        RhythmicEscalationBranches* out) {
    rhythmic_escalation_branches_clear(out);
    if (!input || !intent || !out) return;
    if (!input->escalation_token || strcmp(input->escalation_token, "more_groove") != 0)
        return;
    if (!musical_intent_has_aspect(intent, INTENT_ASPECT_RHYTHMIC))
        return;
    rhythmic_escalation_branches_set_more_groove(out);
}

IntentResult* intent_engine_process(const AbstractInput* input, const ConstraintSet* constraints) {
    if (!input || is_empty_input(input)) {
        MusicalIntent* empty = musical_intent_create();
        if (!empty) return NULL;
        ValueStateSet states = value_state_set_create();
        ViolationList* violations = violation_list_create();
        HarmonicEscalationBranches h_esc;
        RhythmicEscalationBranches r_esc;
        harmonic_escalation_branches_clear(&h_esc);
        rhythmic_escalation_branches_clear(&r_esc);
        IntentResult* result = intent_result_create(empty, &states, violations, "abstain", &h_esc, &r_esc);
        musical_intent_destroy(empty);
        if (violations) violation_list_destroy(violations);
        return result;
    }
    MusicalIntent* intent = infer_intent(input);
    if (!intent) return NULL;
    ValueStateSet states;
    ViolationList* violations = violation_list_create();
    if (!violations) { musical_intent_destroy(intent); return NULL; }
    derive_value_states(intent, constraints, &states, violations);
    HarmonicEscalationBranches h_esc;
    RhythmicEscalationBranches r_esc;
    compute_harmonic_escalation(input, intent, &h_esc);
    compute_rhythmic_escalation(input, intent, &r_esc);
    IntentResult* result = intent_result_create(intent, &states, violations, "processed", &h_esc, &r_esc);
    musical_intent_destroy(intent);
    violation_list_destroy(violations);
    if (result && !validate_invariants(result)) {
        return result;
    }
    return result;
}

File: rust/src/intent_engine.rs
use crate::musical_intent::{IntentAspect, IntentComponent, MusicalIntent};
use crate::value_state::{HarmonicEscalationBranches, RhythmicEscalationBranches, ValueState, ValueStateSet, ValueType};
use crate::constraints::{constraint_overrides_intent, constraints_apply, ConstraintSet, ViolationList};
use crate::intent_result::IntentResult;

#[derive(Debug, Clone)]
pub struct AbstractInput {
    pub input_type: String,
    pub aspect: IntentAspect,
    pub magnitude: f64,
    /// e.g. "more_harmonic_color" — only affects harmony
    pub escalation_token: Option<String>,
}

fn aspect_from_input_type(input_type: &str) -> IntentAspect {
    match input_type {
        "chord_change" | "harmonic" => IntentAspect::Harmonic,
        "rhythmic" => IntentAspect::Rhythmic,
        "dynamic" => IntentAspect::Dynamic,
        "tempo" => IntentAspect::Tempo,
        _ => IntentAspect::Harmonic,
    }
}

pub fn infer_intent(input: &AbstractInput) -> MusicalIntent {
    let aspect = if input.input_type.is_empty() {
        input.aspect
    } else {
        aspect_from_input_type(&input.input_type)
    };
    MusicalIntent::with_components([IntentComponent {
        aspect,
        magnitude: input.magnitude,
        reason: "inferred".to_string(),
    }])
}

pub fn derive_value_states(
    intent: &MusicalIntent,
    constraints: Option<&ConstraintSet>,
    out_states: &mut ValueStateSet,
    out_violations: &mut ViolationList,
) {
    *out_states = ValueStateSet::new();

    let value_types = [
        (ValueType::TempoBias, IntentAspect::Tempo),
        (ValueType::HarmonicTension, IntentAspect::Harmonic),
        (ValueType::RhythmicDensity, IntentAspect::Rhythmic),
        (ValueType::DynamicRange, IntentAspect::Dynamic),
    ];

    for (vt, aspect) in value_types {
        let has_intent = intent.has_aspect(aspect);
        let constrained = constraints
            .map(|c| c.iter().any(|constraint| constraint_overrides_intent(constraint, vt)))
            .unwrap_or(false);

        let state = if !has_intent {
            ValueState::missing("no_intent")
        } else {
            let mag = intent.get_magnitude(aspect);
            ValueState::derive(vt, mag, constrained)
        };
        out_states.set(vt, state);
    }

    if let Some(c) = constraints {
        constraints_apply(c, out_states, out_violations);
    }
}

pub fn validate_invariants(result: &IntentResult) -> bool {
    if !result.harmonic_escalation().satisfies_invariant() {
        return false;
    }
    if !result.rhythmic_escalation().satisfies_invariant() {
        return false;
    }
    for vt in [
        ValueType::TempoBias,
        ValueType::HarmonicTension,
        ValueType::RhythmicDensity,
        ValueType::DynamicRange,
    ] {
        let state = result.value_states().get(vt);
        if !state.validate(vt) {
            return false;
        }
    }
    true
}

fn is_empty_input(input: &AbstractInput) -> bool {
    input.input_type.is_empty()
}

/// Escalation v0: "more_harmonic_color" only. Two mutually exclusive branches: 7ths OR one inversion.
fn compute_harmonic_escalation(input: &AbstractInput, intent: &MusicalIntent) -> HarmonicEscalationBranches {
    if input.escalation_token.as_deref() != Some("more_harmonic_color") {
        return HarmonicEscalationBranches::empty();
    }
    if !intent.has_aspect(IntentAspect::Harmonic) {
        return HarmonicEscalationBranches::empty();
    }
    HarmonicEscalationBranches::more_harmonic_color()
}

/// Escalation v1: "more_groove" only. Three branches: forward push, backward layback, neutral.
fn compute_rhythmic_escalation(input: &AbstractInput, intent: &MusicalIntent) -> RhythmicEscalationBranches {
    if input.escalation_token.as_deref() != Some("more_groove") {
        return RhythmicEscalationBranches::empty();
    }
    if !intent.has_aspect(IntentAspect::Rhythmic) {
        return RhythmicEscalationBranches::empty();
    }
    RhythmicEscalationBranches::more_groove()
}

pub fn intent_engine_process(
    input: &AbstractInput,
    constraints: Option<&ConstraintSet>,
) -> IntentResult {
    if is_empty_input(input) {
        return IntentResult::new(
            MusicalIntent::new(),
            ValueStateSet::new(),
            ViolationList::new(),
            "abstain",
            HarmonicEscalationBranches::empty(),
            RhythmicEscalationBranches::empty(),
        );
    }

    let intent = infer_intent(input);
    let mut states = ValueStateSet::new();
    let mut violations = ViolationList::new();
    derive_value_states(&intent, constraints, &mut states, &mut violations);
    let h_esc = compute_harmonic_escalation(input, &intent);
    let r_esc = compute_rhythmic_escalation(input, &intent);

    IntentResult::new(intent, states, violations, "processed", h_esc, r_esc)
}
